src/freenet/client/async/KeyListenerTracker.java

```diff
@@ -95,61 +102,197 @@ protected static int fixRetryCount(int retryCount) {
 		return Math.max(0, retryCount-MIN_RETRY_COUNT);
 	}
 
+	private boolean contains(KeyListener[] listeners, KeyListener listener) {
+		for(KeyListener l : listeners) {
+			if(l == listener) return true;
+		}
+		return false;
+	}
+
 	public void addPendingKeys(KeyListener listener) {
 		if(listener == null) throw new NullPointerException();
+		byte[] wantedKey = listener.getWantedKey();
+		ByteArrayWrapper wrapper = wantedKey != null ? new ByteArrayWrapper(saltKey(wantedKey)) : null;
+		assert(Arrays.equals(wantedKey, listener.getHasKeyListener().getWantedKey()));
 		synchronized (this) {
 			// We have to register before checking the disk, so it may well get registered twice.
-			if(keyListeners.contains(listener))
-				return;
-			keyListeners.add(listener);
+			if(wantedKey != null) {
+				Object o = singleKeyListeners.get(wrapper);
+				if(o == null) {
+					singleKeyListeners.put(wrapper, listener);
+				} else if(o instanceof KeyListener) {
+					if(listener == (KeyListener)o) return;
```

Comment:
`.equals()` should be used here instead of comparing `KeyListener`s by identity.